File: app/src/main/java/com/example/tanyayuferova/lifestylenews/ui/adapter/ArticlesAdapter.java
package com.example.tanyayuferova.lifestylenews.ui.adapter;

import android.support.v7.widget.RecyclerView;
import android.text.method.LinkMovementMethod;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.tanyayuferova.lifestylenews.databinding.ArticleItemBinding;
import com.example.tanyayuferova.lifestylenews.entity.Article;

import java.util.List;

/**
 * Adapter for displaying articles
 * Created by Tanya Yuferova on 12/17/2017.
 */

public class ArticlesAdapter extends RecyclerView.Adapter<ArticlesAdapter.ArticlesAdapterViewHolder>  {

    public interface OnClickArticleHandler {
        void onClickArticle(View view, Article article);
    }

    private List<Article> data;
    private OnClickArticleHandler clickHandler;

    public ArticlesAdapter() {
    }

    public ArticlesAdapter(OnClickArticleHandler clickHandler) {
        this.clickHandler = clickHandler;
    }

    public void setClickHandler(OnClickArticleHandler clickHandler) {
        this.clickHandler = clickHandler;
    }

    public class ArticlesAdapterViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener{
        private final ArticleItemBinding binding;

        public ArticlesAdapterViewHolder(ArticleItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(Article item) {
            binding.setArticle(item);
            binding.getRoot().setOnClickListener(this);
            binding.tvSourcePublishedOnDate.setMovementMethod(LinkMovementMethod.getInstance());
        }

        @Override
        public void onClick(View v) {
            clickHandler.onClickArticle(binding.getRoot(), binding.getArticle());
        }
    }

    @Override
    public ArticlesAdapterViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        ArticleItemBinding itemBinding = ArticleItemBinding.inflate(layoutInflater, parent, false);
        return new ArticlesAdapterViewHolder(itemBinding);
    }

    @Override
    public void onBindViewHolder(ArticlesAdapterViewHolder holder, int position) {
        Article item = data.get(position);
        holder.bind(item);
    }

    @Override
    public int getItemCount() {
        return data==null ? 0 : data.size();
    }

    public void setData(List<Article> data) {
        this.data = data;
        notifyDataSetChanged();
    }

    public List<Article> getData() {
        return data;
    }
}

File: app/src/main/kotlin/com/tanyayuferova/lifestylenews/ui/list/ArticlesAdapter.kt
package com.tanyayuferova.lifestylenews.ui.list

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.tanyayuferova.lifestylenews.databinding.ItemArticleBinding

/**
 * Author: Tanya Yuferova
 * Date: 7/29/19
 */
class ArticlesAdapter(
    private val actionsHandler: ActionsHandler
) : ListAdapter<ArticleListItem, ArticlesAdapter.ViewHolder>(
    object : DiffUtil.ItemCallback<ArticleListItem>() {
        override fun areItemsTheSame(old: ArticleListItem, new: ArticleListItem): Boolean {
            return old.id == new.id
        }

        override fun areContentsTheSame(old: ArticleListItem, new: ArticleListItem): Boolean {
            return old == new
        }

        override fun getChangePayload(old: ArticleListItem, new: ArticleListItem): Any? {
            return if (old.isFavorite != new.isFavorite)
                IS_FAVORITE_PAYLOAD
            else super.getChangePayload(old, new)
        }
    }
) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val itemBinding = ItemArticleBinding.inflate(layoutInflater, parent, false)
        return ViewHolder(itemBinding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int, payloads: MutableList<Any>) {
        if (payloads.contains(IS_FAVORITE_PAYLOAD)) {
            holder.bindFavorite(getItem(position).isFavorite)
        } else {
            onBindViewHolder(holder, position)
        }
    }

    inner class ViewHolder(
        private val binding: ItemArticleBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(item: ArticleListItem) {
            binding.item = item
            binding.handler = actionsHandler
            bindFavorite(item.isFavorite)
        }

        fun bindFavorite(isFavorite: Boolean) {
            binding.isFavorite = isFavorite
        }
    }

    interface ActionsHandler {
        fun onArticleClick(id: Int)
        fun onFavoriteClick(id: Int, isFavorite: Boolean)
        fun onReadClick(id: Int)
    }

    companion object {
        const val IS_FAVORITE_PAYLOAD = "IS_FAVORITE_PAYLOAD"
    }
}